base/abstractarray.jl

```diff
@@ -495,42 +492,213 @@ function circshift{T,N}(a::AbstractArray{T,N}, shiftamts)
-
-function getindex(A::AbstractVector, i1,i2,i3...)
-    if i2*prod(i3) != 1
-        throw(BoundsError())
+## Approach:
+# We only define one fallback method on getindex for all argument types.
+# That dispatches to an (inlined) internal _getindex function, where the goal is
+# to transform the indices such that we can call the only getindex method that
+# we require AbstractArray subtypes must define, either:
+#       getindex(::T, ::Int) # if linearindexing(T) == LinearFast()
+#       getindex(::T, ::Int, ::Int, #=...ndims(A) indices...=#) if LinearSlow()
+# Unfortunately, it is currently impossible to express the latter method for
+# arbitrary dimensionalities. We could get around that with ::CartesianIndex{N},
+# but that isn't as obvious and would require that the function be inlined to
+# avoid allocations.  If the subtype hasn't defined those methods, it goes back
+# to the _getindex function where an error is thrown to prevent stack overflows.
+#
+# We use the same scheme for unsafe_getindex, with the exception that we can
+# fallback to the safe version if the subtype hasn't defined the required
+# unsafe method.
+
+function getindex(A::AbstractArray, I...)
+    @_inline_meta
+    _getindex(linearindexing(A), A, I...)
+end
+function unsafe_getindex(A::AbstractArray, I...)
+    @_inline_meta
+    _unsafe_getindex(linearindexing(A), A, I...)
+end
+## Internal defitions
+_getindex(::LinearFast, A::AbstractArray) = (@_inline_meta; getindex(A, 1))
+_getindex(::LinearSlow, A::AbstractArray) = (@_inline_meta; _getindex(A, 1))
+_unsafe_getindex(::LinearFast, A::AbstractArray) = (@_inline_meta; unsafe_getindex(A, 1))
+_unsafe_getindex(::LinearSlow, A::AbstractArray) = (@_inline_meta; _unsafe_getindex(A, 1))
+_getindex(::LinearIndexing, A::AbstractArray, I...) = error("indexing $(typeof(A)) with types $(typeof(I)) is not supported")
+_unsafe_getindex(::LinearIndexing, A::AbstractArray, I...) = error("indexing $(typeof(A)) with types $(typeof(I)) is not supported")
+
+## LinearFast Scalar indexing
+_getindex(::LinearFast, A::AbstractArray, I::Int) = error("indexing not defined for ", typeof(A))
+@generated function _getindex(::LinearFast, A::AbstractArray, I::Real...)
+    N = length(I)
+    Isplat = Expr[:(to_index(I[$d])) for d = 1:N]
+    quote
+        $(Expr(:meta, :inline))
+        # We must check bounds for sub2ind; so we can then call unsafe_getindex
+        checkbounds(A, I...)
+        unsafe_getindex(A, sub2ind(size(A), $(Isplat...)))
+    end
+end
+_unsafe_getindex(::LinearFast, A::AbstractArray, I::Int) = (@_inline_meta; getindex(A, I))
+@generated function _unsafe_getindex(::LinearFast, A::AbstractArray, I::Real...)
+    N = length(I)
+    Isplat = Expr[:(to_index(I[$d])) for d = 1:N]
+    quote
+        $(Expr(:meta, :inline))
+        unsafe_getindex(A, sub2ind(size(A), $(Isplat...)))
+    end
+end
+
+# LinearSlow Scalar indexing
+@generated function _getindex{T,AN}(::LinearSlow, A::AbstractArray{T,AN}, I::Real...)
```

Comment:
I definitely agree, but I want to hear more (from you or @tlycken) about how Interpolations would like to use Dual or other non-Real indices. I think these definitions are correct, but that we'll want to change the nonscalar `Union(Real, AbstractVector, Colon)` to be more accepting. 

Does it make sense to use linear indexing with non-real numbers? They would still need to implement arithmetic and div/rem for ind2sub and sub2ind, so perhaps these definitions could be widened to `Number`, but I hesitate to go further. 

For nonscalar indexing, right now all indices must support Base.index_shape, index_lengths, and unsafe_getindex. The tricky one here is indexing - we could have the shape and length default to 1. We could also make to_index default to being a no-op for unknown types.